src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/RazorDynamicFileInfoProvider.cs

```diff
@@ -67,26 +68,46 @@ public RazorDynamicFileInfoProvider(Lazy<RazorWorkspaceListenerInitializer> razo
         Contract.ThrowIfNull(LanguageServerHost.Instance, "We don't have an LSP channel yet to send this request through.");
         var clientLanguageServerManager = LanguageServerHost.Instance.GetRequiredLspService<IClientLanguageServerManager>();
 
-        var response = await clientLanguageServerManager.SendRequestAsync<ProvideDynamicFileParams, ProvideDynamicFileResponse>(
+        var response = await clientLanguageServerManager.SendRequestAsync<RazorProvideDynamicFileParams, RazorProvideDynamicFileResponse>(
```

Comment:
does Razor also use source texts on the other side?  Wondering if it would be worht sending over the Roslyn base document source text checksum over so Razor could verify its returning edits using the same source text as a base

Reply:
@davidwengier I think we do but want to confirm

Reply:
In theory, yes, we use SourceText on the Razor LSP server side. I'm not if the checksum algorithm is in sync, though. Roslyn would presumably use whatever is configured for the project, Razor might not know that. Or maybe the Razor compiler needs it, so it is synced? Worth testing 🤷‍♂️

Reply:
❓ since this is pure text, is there a reason we can't use something like a sha256 checksum for the text on both sides?

Reply:
Other than losing some benefit from Roslyn having presumably already calculated the checksum, no, and I've no idea if that is expensive. I guess the follow up question is what do we do if it fails? Maybe we need to experiment here and see if we have a reliability issue, or if version number is sufficient, or if we need to do something else. I don't think I have a handle on how well this is going to work, to understand how much effort we should put into battening down the hatches

Reply:
My plan was to do the following: 

1. Send version/checksum/whatever in request
2. Check version in client side: 
   * If matches, send diff edits.
   * If doesn't match. Send back full content.
   * Somehow know if it matches and catch in telemetry

Reply:
If we're not sure that the source text is created in the same way (same encoding, same checksum algorithm) there is definitely potential for false negatives.  

However if the source texts have different encodings it seems like the text diff might not be useful anyway (so maybe not a true false negative).  At the very least we'd know when stuff is going off the rails (whereas currently we'd get mangled documents instead)

Reply:
It looks like `RazorCodeDocument.Source.Text` is a `SourceText` that should match given we make sure the encoding and checksum algo are the same. What if we send that data from razor so when we create a sourcetext it's the same? If it's not the same encoding, checksum algo, AND checksum, then we send all content and log telemetry so we know we hit a case that is unexpected. 

That might also cover if a document is opened then closed again, but I'm not sure tbh